src/main/kotlin/com/openlattice/data/ids/PostgresEntityKeyIdService.kt

```diff
@@ -82,15 +88,53 @@ class PostgresEntityKeyIdService(
         return entityKeyIds
     }
 
-    override fun reserveEntityKeyIds( entityKeys: Set<EntityKey> ): Set<UUID> {
-        val entityIdsByEntitySet = entityKeys.groupBy ({ it.entitySetId },{it.entityId}).mapValues { it.value.toSet() }
+    private fun countUpEntityKeys(entityKeys: Set<EntityKey>) {
+        idRefCounts.executeOnKeys(entityKeys, IdRefCountIncrementer())
+
+    }
+
+    private fun countDownEntityKeys(entityKeys: Set<EntityKey>) {
+        idRefCounts.executeOnKeys(entityKeys, IdRefCountDecrementer())
+                .forEach { entityKey, count ->
+                    if (count == 0) {
+                        idMap.delete(entityKey)
+                    }
+                }
+    }
+
+    private fun assignEntityKeyIds(entityKeys: Set<EntityKey>): Map<EntityKey, UUID> {
+        executor.submit {
+            idGenerationService.getNextIds(entityKeys.size).forEach(q::put)
+        }
+
+        return entityKeys.map { key ->
+            val elem = q.take()
+            val assignedId = idMap.putIfAbsent(key, elem)
+            return@map if (assignedId == null) {
+                val assignedPair = key to elem
+                storeEntityKeyIds(mapOf(assignedPair))
+                assignedPair
+            } else {
+                q.put(elem)
```

Comment:
Maybe I'm missing something here, but why do we put back the existing id into the queue?

Reply:
At the time we generate an id we are not guaranteed that it will be assigned, because another node made assign it first (hence the put if absent). If that is not the assigned entity key id then it is returned to the pool of entity key from which keys can be drawn. 

This does make me realize there can be an issue as queue is fixed size and may block on accepting more elements.